### DISCUSSION TITLE
How to resolve ISP hostname or extract ban ip from log

### DISCUSSION THREAD

**Discussion by bwmtx** [fail2ban/fail2ban#3312] [Q&A] [answered]
Looking for a solution to my problem of fail2ban inability to resolve ISP dns name.

My failregex...

```
failregex = ^%(__prefix_line)s.*Login attempt or request with invalid authentication from <HOST>.*$
```
The above works fine except for when I test from my home using my ISP connection.

The error I'm getting...
```
2022-07-01 15:27:04,700 fail2ban.ipdns [1818090]: WARNING Unable to find a corresponding IP address for 35-88-212-11.cnrocmta01.res.dyn.suddenlink.net: [Errno -2] Name or service not known
```

Logfile entry...
```
Login attempt or request with invalid authentication from 35-88-212-11.cnrocmta01.res.dyn.suddenlink.net (35.88.212.11). (Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36)
```

Is there a way to resolve the dns name to an ip or possibly extract the ip address to use for the ban?

Thanks in advance for any help

**Comment by sebres** [comment, marked as answer]
Try to match the IP instead of hostname (and you can use `<ADDR>` tag instead of `<HOST>` to match IPs only).
```diff
- failregex = ... from <HOST>
+ failregex = ... from \S+ \(<ADDR>\)
```
And last but not least - don't use catch-alls (`.*`), because they loosen the anchors and so make the RE vulnerable. Write the RE as precise as possible instead.

**Comment by bwmtx** [comment]
@sebres 

I was just in the process of reading various docs and discovered exactly what you said.  Thanks!

I changed the failregex to your suggestion and made it as precise as possible.  

F2B is picking up the ip and says it's banning, but I still have access?

```
[test] Found 55.233.12.78 - 2022-07-05 11:03:43
[test] Found 55.233.12.78 - 2022-07-05 11:03:45
[test] Ban 55.233.12.78
[test] 55.233.12.78 already banned
```

**Comment by bwmtx** [comment]
@sebres 

Found this...https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-works

I'll see what I can discover.

**Comment by bwmtx** [comment]
@sebres 

So the ban works properly when on the local network.  IP address shows in the iptables chain f2b-test.  When coming from the wan side via reverse proxy, the ban doesn't work.  The only difference I see is, when on local network the source address is an ip address and when coming from wan the source is a domain name.

Output of fail2ban.log
```
[test] Found 35.88.212.11 - 2022-07-05 11:03:43
[test] Found 35.88.212.11 - 2022-07-05 11:03:45
[test] Ban 35.88.212.11
[test] Found 35.88.212.11 - 2022-07-05 11:03:52
[test] Found 35.88.212.11 - 2022-07-05 11:03:58
[test] 35.88.212.11 already banned
```
Output of iptables --list
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
f2b-test     tcp  --  anywhere             anywhere             multiport dports 0:65535

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain f2b-test (1 references)
target     prot opt source               destination         
REJECT     all  --  35-88-212-11.cnrocmta01.res.dyn.suddenlink.net  anywhere             reject-with icmp-port-unreachable
RETURN     all  --  anywhere             anywhere  
```

Maybe the issue the 'reject-with icmp-port-unreachable'?  Also, nslookup doesn't resolve the dns name to an ip address, is it possible to write the ip address, instead of the dns name, to the iptables rule?


/etc/fail2ban/jail.d/test.local
```
[DEFAULT]

[test]
enabled = true
filter = test
logpath = /home/synaptic/test.log
port = 3000
bantime = 30
maxretry = 2
findtime = 86400  
```

**Comment by sebres** [comment]
> when coming from wan the source is a domain name.

What do you mean? The FQDN in listing of `iptables --list`? No, it is not - you need to use `-n` to get numeric values (IP addresses, ports etc).

> Maybe the issue the 'reject-with icmp-port-unreachable'?

No.
Although I would not use multiport 0:65535, but allports action instead. Just it does not really make a difference (both must ban the IP for every port).

> is it possible to write the ip address, instead of the dns name, to the iptables rule?

They are always written as an IPs. It will be resolved in iptables listing if called without `-n`.

I can imagine that this is either some keep-alive connection (an some rules in firewall avoid a reject of established connections).
Or it is something other wrong - e. g. it is not TCP proto (UDP traffic for instance).
Or you're trying to ban wrong IP address (e. g. from the log-file it is something like `X-Forwarded-For` header, so the IP behind the proxy, but the request is coming from proxying side - which is surely different IP address).
In this case you have to ban it either on the proxy side (action blocking the IP there) or use different banning action like `nginx-block-map` allowing to block inside the proxy by the header (see https://github.com/fail2ban/fail2ban/issues/3256#issuecomment-1100033154 for similar question).

**Reply by bwmtx** [reply]
TCP.  Yes, I am using the ```X-Forwarded-For``` header in Nginx config.  

The connection going over a reverse ssh tunnel ```ssh -NT -R 60000:localhost:3000 -i /home/synaptic/.ssh/id_ed25519 synaptic@host``` to cloud server, cloud server proxies connection to domain name over the reverse tunnel.

I checked the nginx access log on the cloud server.  It has no idea of the failed login(s).

With the above said, would I need to ban the cloud server ip or localhost?  Either way I need to specify a ban ip other than the detected ip.   Is this the correct approach?

Would it make more sense to run a script when a ban happens and write it to a log file on the cloud server so fail2ban can act on it?

Appreciate you thoughts / assistance!

**Reply by sebres** [reply]
> With the above said, would I need to ban the cloud server ip or localhost?

No. This is a nonsense (then it would ban everything, at least a lot of users going through the proxy).
You need surely to ban only the real IPs of evildoers, but as already said it is impossible with net-filter/firewall banning actions like iptables on the side of fail2ban (where the connection is initiated from proxy and not from evil IP). 
Either you need some action traversing the IP to proxying side (same side where IP initially arrives) in order to ban the IP there, or you need some special action (like nginx-block-map) banning the connection by the header.

Assume the IP of evildoer is `IP-E`, the IP of proxy-side is `IP-P`, the IP of your host with fail2ban is `IP-H`.
Then the chain looks like this:
```
           |   (1)    |                |       (2)        |
[IP-E] --> | Internet | --> [IP-P] --> | WAN/NAT/Intranet | --> [IP-H]
           |          |                |                  |
```
Banning of `IP-E` with firewall/net-filter is only possible on side of host `IP-P`, because only the connection (1) has this `IP-E` as source IP and connection (2) has `IP-P` as source IP. However on side of `IP-H` you can only reject the request in the web-server or its backend by the header that contains `IP-E`, whereas net-filter sees there only `IP-P` as source. Also note that a reject of whole connection (2) may be wrong due to many reasons, e. g. it can be used to serve multiple requests (multiplexing), as well as too error-prone or even impossible (`IP-E` is not visible there due to encryption e. g. https-protocol).
Again, to ban `IP-E` either you'd somehow send the banning info to `IP-P` with some custom action and it must initiate the iptables rules there (create/delete them by ban/unban) if this side is under your control, or you'd ban `IP-E` by the header directly in web-server or backend on `IP-H` with action like [nginx-block-map](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/nginx-block-map.conf).
There are no other variants (excepting possibly some gateways like cloudflare if connection (1) takes place over that and the gateway has an interface to ban the IP there, so you'd also need some action sending the ban/unban tickets to that gateway).

**Comment by bwmtx** [comment]
@sebres 

I can't use nginx-block-map as the proxy [IP-P] has no idea of the login failures on [IP-H].  Only thing I can think of is running a banaction script on host [IP-H] to send ban details to [IP-P] (write to a log file that fail2ban can read).

Explain, or point me to docs, how to run a bash script for ban action passing the ip as an argument?

**Comment by sebres** [comment]
> I can't use nginx-block-map as the proxy [IP-P] has no idea of the login failures on [IP-H]. 

Well, that was just an example and you can either write your own custom action doing it similarly to nginx-block-map but for your backend (so rejecting the request inside the backend), or even to distribute the log to `IP-P` side (and monitor it there with another fail2ban instance).
Also it is possible on `IP-P` side if your backend (on `IP-H` side) would reject the request with some special response code, e. g. 400, so the  nginx on `IP-P` side would log it and then a fail2ban instance there monitoring that error or access log, could find this and ban the IP.

There are many many possibilities how one can do that. Just note [wiki :: Best practice # Reduce parasitic log-traffic](https://github.com/fail2ban/fail2ban/wiki/Best-practice#reduce-parasitic-log-traffic) if you'd decide to monitor access log.

> Explain, or point me to docs, how to run a bash script for ban action passing the ip as an argument?

Hmm... Simply take a look into some action of us, for example a [dummy](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/dummy.conf) action (basically every non-pythonic action is a command action). 
It must have at least `actionban`, `actionunban` (and probably `actionflush` to clean up the IPs in bulk mode by shutdown).

The simplest action may look like this:

```ini
[Definition]

_send2proxy = ssh user@host /some/path/proxy-firewall.sh

actionflush = <_send2proxy> flush
actionban = <_send2proxy> ban <ip> family <family>
actionunban = <_send2proxy> unban <ip> family <family>
```

This way you could also try to implement the proxying action for some of already available net-filter actions of us (e. g. [iptables-ipset](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/iptables-ipset.conf)):

```ini
[INCLUDES]

before = iptables-ipset.conf

[Definition]

_send2proxy = ssh user@host

actioncheck = <_send2proxy> "<known/actioncheck>"
actionstart = <_send2proxy> "<known/actionstart>"
actionstop = <_send2proxy> "<known/actionstop>"
actionflush = <_send2proxy> "<known/actionflush>"
actionban = <_send2proxy> "<known/actionban>"
actionunban = <_send2proxy> "<known/actionunban>"
```
(just you must ensure that the original action parameters doesn't use quote char `"` inside, otherwise you need to escape or somehow modify the command scripts supplying all that to the ssh with a single script parameter)

As for the documentation, see man of [jail.conf.5](https://manpages.debian.org/testing/fail2ban/jail.conf.5.en.html#ACTION_CONFIGURATION_FILES_(action.d/*.conf))

**Reply by bwmtx** [reply]
##########
On IP-H
##########

jail.d/test.local
```
[test]
enabled = true
filter = test
logpath = /path/to/logfile.log
maxretry = 2
findtime = 86400
```
My understanding after review of man jail.conf is that by default it will load action.d/test.local.  Correct?

action.d/test.local
```
[Definition]
actionban = ./etc/fail2ban/test.sh <ip>
```

test.sh
```
#!/bin/bash

IP=$1
MESSAGE="Ban $IP"

<Send $MESSAGE to proxy server log file which fail2ban uses to ban $IP>
```

##########
On IP-P
########## 

filter.d/test.local
```
[INCLUDES]
before = common.conf

[Definition]
failregex = ^Ban <ADDR>$
```

jail.d/test.local
```
[test]
enabled = true
filter = test
logpath = /path/to/logfile.log
bantime = 3600
```

**Comment by sebres** [comment]
> My understanding after review of man jail.conf is that by default it will load action.d/test.local. Correct?

No. Do you mean `filter = test` doing that? No, in that case it loads monitoring rules like `failregex` etc from `filter.d/test.local`.
To specify an action you need to set `banaction` (or `action` which interpolate `banaction` internally) parameter. Otherwise it uses an action defined in default section.
https://github.com/fail2ban/fail2ban/blob/3c22ae161cfe23a9fb00cabb8a7de891e3e8edaf/config/jail.conf#L208-L212
and
https://github.com/fail2ban/fail2ban/blob/3c22ae161cfe23a9fb00cabb8a7de891e3e8edaf/config/jail.conf#L263

So I guess you'd do this either:
```diff
 [test]
 enabled = true
 filter = test
+banaction = test
 ...
```
The issue with this approach would be that the jails `IP-H` and `IP-P` must still get the same `bantime` (otherwise it could be ignored by later attempts on `IP-H` with "already banned") as well as the jail on `IP-P` must get `maxretry = 1`, because I guess it should ban directly by first message from `IP-H`.

Also you need to write a timestamp to the log for `IP-P`, because fail2ban expects it by default, so ensure it is written with `test.sh`.

As for `filter = test` you don't need it for such simple case, you can set `filter` to empty and specify `failregex` directly in the jail.
So for the jail on `IP-P` it would be:
```ini
[test]
enabled = true
filter =
failregex = ^\s*Ban <ADDR>
maxretry = 1
...
````

PoC:
```
$ fail2ban-regex "$(date +%s) Ban 192.0.2.123" "^\s*Ban <ADDR>"
...
Lines: 1 lines, 0 ignored, 1 matched, 0 missed
```

**Reply by bwmtx** [reply]
> > My understanding after review of man jail.conf is that by default it will load action.d/test.local. Correct?
> 
> No. Do you mean `filter = test` doing that? No, in that case it loads monitoring rules like `failregex` etc from `filter.d/test.local`. To specify an action you need to set `banaction` (or `action` which interpolate `banaction` internally) parameter. Otherwise it uses an action defined in default section.
> 
> https://github.com/fail2ban/fail2ban/blob/3c22ae161cfe23a9fb00cabb8a7de891e3e8edaf/config/jail.conf#L208-L212
> 
> 
> and
> https://github.com/fail2ban/fail2ban/blob/3c22ae161cfe23a9fb00cabb8a7de891e3e8edaf/config/jail.conf#L263
> 
> So I guess you'd do this either:
> 
> ```diff
>  [test]
>  enabled = true
>  filter = test
> +banaction = test
>  ...
> ```
> 
> The issue with this approach would be that the jails `IP-H` and `IP-P` must still get the same `bantime` (otherwise it could be ignored by later attempts on `IP-H` with "already banned") as well as the jail on `IP-P` must get `maxretry = 1`, because I guess it should ban directly by first message from `IP-H`.
> 
> Also you need to write a timestamp to the log for `IP-P`, because fail2ban expects it by default, so ensure it is written with `test.sh`.
> 
> As for `filter = test` you don't need it for such simple case, you can set `filter` to empty and specify `failregex` directly in the jail. So for the jail on `IP-P` it would be:
> 
> ```ini
> [test]
> enabled = true
> filter =
> failregex = ^\s*Ban <ADDR>
> maxretry = 1
> ...
> ```
> 
> PoC:
> 
> ```
> $ fail2ban-regex "$(date +%s) Ban 192.0.2.123" "^\s*Ban <ADDR>"
> ...
> Lines: 1 lines, 0 ignored, 1 matched, 0 missed
> ```

**Comment by bwmtx** [comment]
Thanks for the pointers and assistance!  I'll try my hand at putting a working solution together.
